<h1>ISAT 449 - Emerging Topics in Applied Data Science</h1>
<h2>TensorFlow-Keras: Single Layer Single Neuron Model</h2>
<h3>Training Your First TensorFlow-Keras Model (Temperature Conversion)</h3>
<h3>Problem and Objectives</h3>
<p>Let's solve the problem from the lecture with TensorFlow and Keras. We want to create a model that converts from degree Celsius to degree Fahrenheit. The formula, that we know from our physics class is:</p>
$F=\frac{9}{5}\times C+32$
<p>We could easily solve this using Python but we want to illustrate how to do it with machine learning.</p>
<ul>
  <li>To this end, we will give TensorFlow some sample Celsius data values (-40,-10, 0, 8, 15, 22, 38) and their corresponding Fahrenheit data values (-40,14, 32, 46, 59, 72, 100).</li>
  <li>Then, we will train a model that figures out the above formula through the training process.</li>  
</ul>


<h1>Imports</h1>
<ul>
  <li>First, import TensorFlow. Here, we're calling it tf (an alias), for ease of use.
  </li>
  <li>Next, import NumPy (http://www.numpy.org/) as np . Numpy is a high performining numerical library that helps us represent and process our data as high performing arrays.
  </li>
</ul>

In [ ]:
#Importing
import tensorflow as tf
import numpy as np
print('TensorFlow version:', tf.__version__)

TensorFlow version: 2.6.0


<h2>Set up training data</h2>
<p>This is Supervised Machine Learning where we wll figure out an algorithm given a set of inputs and outputs. Since the task in this Colab is to create a model that can give the
temperature in Fahrenheit when given the degrees in Celsius, we create two lists, celsius and fahrenheit , that we can use to train our model.
</p>

In [ ]:
#create numpy arrays from python lists
celsius = np.array([-40,-10,0,8,15,22,38], dtype=float)
fahrenheit = np.array([-40,14,32,46,59,72,100],dtype=float)

#print the data pairs
for i,c in enumerate(celsius):
  print('{} degrees Celsius = {} degrees Fahrenheit'.format(c, fahrenheit[i]))

-40.0 degrees Celsius = -40.0 degrees Fahrenheit
-10.0 degrees Celsius = 14.0 degrees Fahrenheit
0.0 degrees Celsius = 32.0 degrees Fahrenheit
8.0 degrees Celsius = 46.0 degrees Fahrenheit
15.0 degrees Celsius = 59.0 degrees Fahrenheit
22.0 degrees Celsius = 72.0 degrees Fahrenheit
38.0 degrees Celsius = 100.0 degrees Fahrenheit


<h2>Recall Important Machine Learning Technology</h2>
<ul>
  <li>Feature(s) — The input(s) to our model. In this case, a single value — the degrees in Celsius.
  </li>
  <li>Labels — The output our model predicts. In this case, a single value — the degrees in Fahrenheit.
  </li>
  <li>Sample — A pair of inputs/outputs used during training. In our case a pair of values from celsius and fahrenheit at a specific index, such as (15,59) .
  </li>
</ul>
<br>
<h2>Create the model</h2>
<p>Next create the model. We will use simplest possible model we can, a Dense network. Since the problem is straightforward, this network will require only a single layer, with a single
neuron.
</p>
<br>
<h2>Build a layer</h2>
<p>We'll call the layer l0 (for layer zero) and create it by instantiating tf.keras.layers.Dense with the following configuration:
<ul>
  <li>input_shape=[1] — This specifies that the input to this layer is a single value. That is, the shape is a one-dimensional array with one member. Since this is the first (and only)
layer, that input shape is the input shape of the entire model. The single value is a floating point number, representing degrees Celsius</li>
  <li>units=1 — This specifies the number of neurons in the layer. The number of neurons defines how many internal variables the layer has to try to learn how to solve the problem
(more later). Since this is the final layer, it is also the size of the model's output — a single float value representing degrees Fahrenheit. (In a multi-layered network, the size and
shape of the layer would need to match the input_shape of the next layer.)</li>
</ul>
</p>
<p>
The code to do this, as discussed in the lecture is:
</p>

In [ ]:
l0 = tf.keras.layers.Dense(units=1, input_shape=(1,))

<h2>Assemble layers into the model</h2>
<h2>The Sequential Model</h2>
<p>Once layers are defined, they need to be assembled into a model. The Sequential model definition takes a list of layers as argument, specifying the calculation order from the input to
the output.
<br>
This model has just a single layer, l0 .
</p>


In [ ]:
model = tf.keras.Sequential([l0])

<h6>Note</h6>
<p>You will often see the layers defined inside the model definition, rather than beforehand</p>

```
model = tf.keras.Sequential([
  tf.keras.layers.Dense(units=1, input_shape=[1])
])
```
<h3>Look at the model's architecture</h3>


In [ ]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 1)                 2         
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________


<h1>Compile the model, with loss and optimizer functions</h1>
<p>Before training, the model has to be compiled. When compiled for training, the model is given:</p>
<ul>
  <li>Loss Function - A way of measuring how far off predictions are from the desired outcome. (The measured difference is called the "loss".</li>
  <li>Optimizer Function - A way of adjusting internal values in order to reduce the loss</li>
</ul>

In [ ]:
model.compile(loss='mean_squared_error',
              optimizer=tf.keras.optimizers.Adam(0.1))

<p>These are used during training ( model.fit() , below) to first calculate the loss at each point, and then improve it. In fact, the act of calculating the current loss of a model and then
improving it is precisely what we mean by training a model.
<br>
During training, the optimizer function is used to calculate adjustments to the model's internal variables. The goal is to adjust the internal variables until the model (which is really a math
function) mirrors the actual equation for converting Celsius to Fahrenheit.
<br>
TensorFlow uses numerical analysis to do this. What is useful to know about these parameters are:
<ul>
  <li>The loss function (mean squared error (https://www.youtube.com/watch?v=mdKjMPmcWjY&ab_channel=CodeEmporium)) used here and the optimizer (Adam
(https://www.wikidata.org/wiki/Q29410287))are standard for simple models like this one, but many others are available.
</li>
  <li>One part of the Optimizer you may need to think about when building your own models is the learning rate ( 0.1 in the code above). This is the step size taken when adjusting
values in the model.
    <ul>
      <li>If the value is too small, it will take too many iterations to train the model</li>
      <li>Too large, and accuracy goes down</li>
      <li>Finding a good value often involves some trial and error, but the range is usually within 0.001 (default), and 0.1
</li>
    </ul>
  </li>
</ul>

<h2>Train the model</h2>
<p>Train the model by calling the fit method</p>
<p>During training, the model takes in Celsius values, performs a calculation using the current internal variables (called "weights") and outputs values which are meant to be the Fahrenheit
equivalent. Since the weights are initially set randomly, the output will not be close to the correct value. The difference between the actual output and the desired output is calculated
using the loss function, and the optimizer function directs how the weights should be adjusted</p>
<p>This cycle of calculate, compare, adjust is controlled by the fit method</p>
<ol>
  <li>The first argument is the inputs</li>
  <li>The second argument is the desired outputs</li>
  <li>The epochs argument specifies how many times this cycle should be run</li>
  <li>The verbose argument controls how much output the method produces</li>
</ol>

In [ ]:
history = model.fit(celsius, fahrenheit, epochs=500, verbose=False)
print('Finished training the model')

Finished training the model


<h2>Use the model to predict values</h2>
<p>Now you have a model that has been trained to learn the relationship between celsius and fahrenheit . You can use the predict method to have it calculate the Fahrenheit
degrees for a previously unknown Celsius degrees.</p>
<br>
<p>So, for example, if the Celsius value is 100, what do you think the Fahrenheit result will be? Take a guess before you run this code.
</p>

In [ ]:
print('Model predicts that the 100 degrees Celsius corresponds to {0:.3f} degrees Fahrenheit'.format(model.predict([100.0])[0][0]))

Model predicts that the 100 degrees Celsius corresponds to 211.329 degrees Fahrenheit


<h3>To review</h3>
<ul>
  <li>We created a model with a Dense Layer</li>
  <li>We trained it with 3500 examples(7 pairs, over 500 epochs)</li>
</ul>
<p>Our model tuned the variables (weights) in the Dense layer until it was able to return the correct Fahrenheit value for any Celsius value. (Remember, 100 Celsius was not part of our
training data.)
</p>
<h2>Looking at the layer weights</h2>
<p>Finally, lets print the internal variables of the Dense layer</p>


In [ ]:
print('These are the layer variables: {}'.format(l0.get_weights()))

These are the layer variables: [array([[1.8218627]], dtype=float32), array([29.142672], dtype=float32)]


<p>The first variable is close to ~1.8 (9/5 = 1.8 and the second to ~32. These values (1.8 and 32) are the actual variables in the real conversion formula.
<br>
This is really close to the values in the conversion formula. We'll explain this in an upcoming video where we show how a Dense layer works, but for a single neuron with a single input
and a single output, the internal math looks the same as the equation for a line (https://en.wikipedia.org/wiki/Linear_equation#Slope%E2%80%93intercept_form),𝑦 = 𝑚𝑥 + 𝑏 , which
has the same form as the conversion equation, .
<br>
Since the form is the same, the variables should converge on the standard values of 1.8 and 32, which is exactly what happened.
𝑓 = 1.8𝑐 + 32
<br>
So based on the weight (slope) and bias(intercept) our model is:
</p>
$$F=1.82\times C+29.24$$
<p>With additional neurons, additional inputs and additional outputs, the formula becomes much more complex, but the idea is the same.</p>
<h3>Comment</h3>
<p>As we note above, in practice, you will hardly ever see the layer(s) of a model defined beforehand and outside of the Sequential statement. What you will see is see the layers
defined inside the model definition as indicated below:</p>


```
model = tf.keras.Sequential([
  tf.keras.layers.Dense(units=1, input_shape=[1])
])
```

<p>You should especially take note that the layer(s) are elements of a python list</p>

<h2>Exercises</h2>
<h4>Exercise 1</h4>
<p>Reconstruct our Sequential model using this appoach and train it on the same dataset. The statements to compile and train (fit) it are exacly the same. Put all of your code here:</p>

In [ ]:
model = tf.keras.Sequential([
  tf.keras.layers.Dense(units=1, input_shape=[1])
])

model.compile(loss='mean_squared_error',optimizer=tf.keras.optimizers.Adam(0.1))

model.fit(celsius, fahrenheit, epochs=500, verbose=False)
print('Done!')

Done!


<h4>Exercise 2</h4>
<p>Add two additional layers to your Sequential model using this appraoch. The second layer should have 3 neurons and the final layer should have 1 neuron so the three layers would be:</p>
<ul>
  <li>l0 = tf.keras.layers.Dense(units=1, input_shape=[1])</li>
  <li>l1 = tf.keras.layers.Dense(units=3)</li>
  <li>l2 = tf.keras.layers.Dense(units=1)</li>
</ul>
<p>You should note that each additional layer that adds neurons (units) also adds more variables to your model and the simple two variable relationship we saw above become far more
complex.
<br>
In the cell below call the summary method of the model and examine its architecture. The code to do this(as we saw above) is :

```
model.summary()
```
Briefly comment inidcating your understanding. Specifically,can you explain the value for Total params ? HINT: Don't forget to count biases.



In [ ]:
model.summary()
#Layer type is basically saying what kind of layer it is, we put dense.
#Output shape is basically the input_shape, we used [1]
#Param # is the fit, so celsius and fahrenheit
#Total params: 2 means that the things we fit based on which is again the celsius and fahrenheit

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_5 (Dense)              (None, 1)                 2         
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________


<h4>Exercise 3</h4>
<p>Run the code below for the model</p>

In [ ]:
#create layers
l0 = tf.keras.layers.Dense(units=4,input_shape=[1])
l1 = tf.keras.layers.Dense(units=4)
l2 = tf.keras.layers.Dense(units=1)

#assemble the model
model = tf.keras.Sequential([l0,l1,l2])

#compile the model
model.compile(loss='mean_squared_error', optimizer=tf.keras.optimizers.Adam(0.1))

#train the model
model.fit(celsius, fahrenheit, epochs=500, verbose=False)

print('Finished training the model\n')
#make a prediction
print('Model predicts that 100 degrees Celsius is: {0:.3f} degrees Fahrenheit\n'.format(model.predict([100])[0][0]))

Finished training the model

Model predicts that 100 degrees Celsius is: 211.747 degrees Fahrenheit



<p>Now, let's examine the weights in each layer. Make sure you add up the number of varaibles to get the total.
</p>

In [ ]:
print('These are the layer zero (l0) variables: \n{}\n'.format(l0.get_weights()))
print('These are the layer one (l1) variables: \n{}\n'.format(l1.get_weights()))
print('There are the layer two (l2) variables: \n()\n'.format(l2.get_weights()))

These are the layer zero (l0) variables: 
[array([[ 0.37408137, -0.39742556, -0.9571982 ,  0.22197145]],
      dtype=float32), array([-2.5699284, -3.323159 , -3.3697538, -0.502955 ], dtype=float32)]

These are the layer one (l1) variables: 
[array([[-0.3427481 ,  0.28380033, -1.2585084 , -0.9952351 ],
       [-1.2022073 ,  0.2173424 , -1.2692256 ,  0.19481821],
       [-1.1225858 , -0.3077973 ,  0.16608585, -0.622954  ],
       [ 0.27415904,  0.5499032 , -0.73264766,  0.1690531 ]],
      dtype=float32), array([3.333244 , 2.448188 , 3.3790123, 3.3851452], dtype=float32)]

There are the layer two (l2) variables: 
()



<p>As you can see, this model is also able to predict the corresponding Fahrenheit value really well. But when you look at the variables (weights) in the l0 and l1 layers, they look
nothing like we previously observerd and are not even close to ~1.8 and ~32. The added complexity hides the "simple" form of the conversion equation.
<br>
Now call the model's summary method: model.summary()
</p>

In [ ]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 4)                 8         
_________________________________________________________________
dense_2 (Dense)              (None, 4)                 20        
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 5         
Total params: 33
Trainable params: 33
Non-trainable params: 0
_________________________________________________________________


<p>Give a short (one sentence!) explanation of why there are eight values in the l0 array</p>
<p></p>